### ISSUE TITLE
record typings broken when domain is a branded type

### ISSUE THREAD

**Issue by brabeji** [gcanti/io-ts#360]
Hello @gcanti, I think we may have found a bug in record type.

### Current Behavior

https://codesandbox.io/s/frosty-ellis-lynyt?module=/src/index.ts

![image](https://user-images.githubusercontent.com/2237954/63755553-34af8800-c8b7-11e9-8829-c390cd15fd8c.png)

![image](https://user-images.githubusercontent.com/2237954/63755534-2b262000-c8b7-11e9-9dc2-6d540504a682.png)

### Expected behavior

The branded type domain should be typed as string.

### Reproducible example

https://codesandbox.io/s/frosty-ellis-lynyt?module=/src/index.ts

### Suggested solution(s)

Fix typings of `record`

### Your environment

| Software   | Version(s) |
| ---------- | ---------- |
| io-ts      |       2.0.1     |
| fp-ts      |      2.0.5      |
| TypeScript |      3.3.3      |

We <3 `io-ts`, thank you!

**Comment by poteat** [comment]
I ran into this as well.  I think your report only concerns itself with type-level behavior, but there is a runtime-type level bug as well:

If you construct a `t.record` such that the domain is a custom branded string type `CustomString`, and the codomain is a `t.number`, the `decode(x: any)` function of that record will not check that all keys of `x` match the custom type guard of `CustomString`.

To put it another way, no logical conditions you define as part of your branded domain type will be checked during runtime.

**Comment by DenisFrezzato** [comment]
This is actually an "issue" with the type system, check this https://github.com/microsoft/TypeScript/issues/15746#issuecomment-301922834.
> The current behavior or Record takes the known literal types in K, which in this case is the empty set and creates a type for you, and the resulting type is {}.
